### PR TITLE
Branch Clause from Predicate

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk 

This is part of what's proposed in https://github.com/rust-lang/compiler-team/issues/531

### CODE REVIEW

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 1]
```diff
@@ -666,15 +666,17 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         let tcx = self.infcx.tcx;
 
         // Find out if the predicates show that the type is a Fn or FnMut
-        let find_fn_kind_from_did =
-            |predicates: ty::EarlyBinder<&[(ty::Predicate<'tcx>, Span)]>, substs| {
-                predicates.0.iter().find_map(|(pred, _)| {
+        let find_fn_kind_from_did = |predicates: ty::EarlyBinder<
+            &[(ty::Predicate<'tcx>, Span)],
+        >,
+                                     substs| {
```
Comment (oli-obk) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:672]: Go home rustfmt, you're drunk

Comment (spastorino) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:672]: Hehehe, we could add a FIXME: re run fmt when bug #XYZ is fixed.

Comment (Nilstrieb) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:672]: This appears to be caused be the let_chain inside the closure, as removing it causes it to be formatted correctly for me.

**Comment by oli-obk** [comment]
The diesel regression is

```
  244,159,238  ???:<rustc_trait_selection::traits::fulfill::FulfillmentContext as rustc_infer::traits::engine::TraitEngine>::register_predicate_obligation
  128,790,199  ???:rustc_trait_selection::traits::wf::predicate_obligations
 -112,762,937  ???:<&mut rustc_hir_analysis::check::wfcheck::check_where_clauses::{closure
   70,951,310  ???:rustc_middle::ty::util::fold_list::<rustc_infer::infer::canonical::canonicalizer::Canonicalizer, rustc_middle::ty::Predicate, <&rustc_middle::ty::list::List<rustc_middle::ty::Predicate> as rustc_middle::ty::fold::TypeFoldable>::try_fold_with<rustc_infer::infer::canonical::canonicalizer::Canonicalizer>::{closure
   37,831,490  ???:<rustc_middle::ty::context::CtxtInterners>::intern_predicate
   24,179,850  ???:<rustc_infer::infer::InferCtxt>::resolve_vars_if_possible::<rustc_infer::traits::Obligation<rustc_middle::ty::Predicate>>
   21,434,192  ???:<rustc_trait_selection::traits::fulfill::FulfillProcessor as rustc_data_structures::obligation_forest::ObligationProcessor>::process_obligation
```

We should figure out what part of this PR caused this

**Comment by spastorino** [comment]
@rustbot label +perf-regression-triaged

We're aware of this regression, and I've opened https://github.com/rust-lang/rust/issues/105060 to track it. We would need to intern Clauses in order to make the memory used lower, comparisons (comparing references) faster and such. After @oli-obk lands some stuff he is doing, I'm going complete the thing by interning Clauses.

**Pull request merged**